File: C++/28. Implement strStr().cpp
class Solution {
public:
    int strStr(string haystack, string needle) {
        int k = -1, i = 0, len = needle.length();
        if(len==0)
            return 0;
        vector<int> nex(len+1,-1);
        while(i<len)
        {
            if(k==-1 || needle[i] == needle[k])
            {
                i++,k++;
                nex[i] = k;
            }
            else
                k = nex[k];
        }
        int j = 0;
        i = 0;
        int len1 = haystack.length(), len2 = needle.length();
        while(i<len1 && j<len2)
        {
            if(haystack[i] == needle[j])
            {
                i++;
                j++;
            }
            else if(j==0)
                i++;
            else
                j = nex[j];
        }
        if(j==len2)
            return i-len2;
        return -1;
    }
};

File: Python/28. Implement strStr().py
class Solution:
    def strStr(self, haystack, needle):
        """
        :type haystack: str
        :type needle: str
        :rtype: int
        """
        nex = [-1] * (len(needle)+1)
        k, i, j = -1, 0, 0
        while i < len(needle):
            if k == -1 or needle[i] == needle[k]:
                i += 1
                k += 1
                nex[i] = k
            else:
                k = nex[k]
        i = 0
        while i < len(haystack) and j < len(needle):
            if haystack[i] == needle[j]:
                i += 1
                j += 1
            elif j == 0:
                i += 1
            else:
                j = nex[j]
        if j == len(needle):
            return i - j
        return -1